src/Tasks/AssemblyDependency/Resolver.cs

```diff
@@ -312,23 +321,47 @@ List<ResolutionSearchLocation> assembliesConsideredAndRejected
             if (directory != null)
             {
                 string weakNameBase = assemblyName.Name;
+
+                // feature flag by ChangeWaves 
+                bool useFileEnumerationOptimization = Utilities.ChangeWaves.AreFeaturesEnabled(Utilities.ChangeWaves.Wave17_0);
+
                 foreach (string executableExtension in executableExtensions)
                 {
-                    string baseName = weakNameBase + executableExtension;
                     string fullPath;
+                    bool fileExistenceVerified = false;
+
+                    string fileName = $"{weakNameBase}{executableExtension}";
+                    if (useFileEnumerationOptimization)
+                    {
+                        try
+                        {
+                            fileExistenceVerified = getDirectoryFile(directory, fileName) != null;
+                        }
+                        catch (Exception e) when (ExceptionHandling.IsIoRelatedException(e))
+                        {
+                            // Assuming it's the search path that's bad. But combine them both so the error is visible if it's the reference itself.
+                            throw new InvalidParameterValueException("SearchPaths", directory + (directory.EndsWith("\\", StringComparison.OrdinalIgnoreCase) ? String.Empty : "\\") + fileName, e.Message);
+                        }
+
+                        if (!fileExistenceVerified)
+                        {
+                            // file do not exists, try next file extension
+                            continue;
+                        }
+                    }
 
                     try
                     {
-                        fullPath = Path.Combine(directory, baseName);
+                        fullPath = Path.Combine(directory, fileName);
                     }
                     catch (Exception e) when (ExceptionHandling.IsIoRelatedException(e))
                     {
                         // Assuming it's the search path that's bad. But combine them both so the error is visible if it's the reference itself.
-                        throw new InvalidParameterValueException("SearchPaths", directory + (directory.EndsWith("\\", StringComparison.OrdinalIgnoreCase) ? String.Empty : "\\") + baseName, e.Message);
+                        throw new InvalidParameterValueException("SearchPaths", directory + (directory.EndsWith("\\", StringComparison.OrdinalIgnoreCase) ? String.Empty : "\\") + fileName, e.Message);
```

Comment:
```suggestion
                        throw new InvalidParameterValueException("SearchPaths", Path.Combine(directory, fileName), e.Message);
```

Reply:
Path.Combine could throw

Reply:
Should still use `Path.DirectorySeparatorChar` instead of the hard-coded backslash so it works well on Linux and Mac as well.